### ISSUE TITLE
Implement missing Optimizers

### ISSUE THREAD

**Issue by louisfd** [burn-rs/burn#543]
Read the section on optimization before trying to implement an optimizer: https://github.com/burn-rs/burn/blob/main/ARCHITECTURE.md#optimization

- [X] SGD
- [X] Adam
- [x] AdamW
- [x] RMSProp
- [x] AdaGrad

**Comment by AuruTus** [comment]
Hi, I'm interested in this and working on `RMSProp`. is there any other guy who has already started on it?

**Comment by antimora** [comment]
@AuruTus go ahead and claim it. @wdoppenberg currently has an open PR https://github.com/burn-rs/burn/pull/566 for AdamW. You can see how it's been done.

**Comment by wdoppenberg** [comment]
> @AuruTus go ahead and claim it. @wdoppenberg currently has an open PR #566 for AdamW. You can see how it's been done.

It's a WIP still, but I'd be happy to collaborate in some way if you're interested. Feel free to message me on discord.

**Comment by CohenAriel** [comment]
Hello. I'd like to work on AdaGrad if there's no one already working on it.

**Comment by antimora** [comment]
> Hello. I'd like to work on AdaGrad if there's no one already working on it.

go ahead ! happy coding.

**Comment by AuruTus** [comment]
hi @CohenAriel , I noticed that the common `weight_decay` 's `transform` method has some bugs, so almost all optimizers will be affected. Can use temporary fix to make a workaround now. I think I can fix it later but there are still some bugs in my RMSProp now 😅.

The bug code and discussion can be refered here on [discord](https://discord.com/channels/1038839012602941528/1091796857996451942/1139895797958332426)

and this is my temp impl for ref:
```rust
    /// temp fix for Transform.
    pub fn transform_temp_fix<const D: usize>(
        &self,
        grad: Tensor<B, D>,
        tensor: Tensor<B, D>,
        state: Option<WeightDecayState<B, D>>,
    ) -> (Tensor<B, D>, Tensor<B, D>, WeightDecayState<B, D>) {
        let grad = match state {
            Some(state) => state.grad_last_step.mul_scalar(self.penalty).add(grad),
            None => tensor.clone().mul_scalar(self.penalty).add(grad),
        };

        (grad, tensor.clone(), WeightDecayState::new(tensor))
    }
```

**Comment by CohenAriel** [comment]
Wow, thanks! I guess I can just not use weight decay in the test until it's fixed

**Comment by AuruTus** [comment]
@CohenAriel @wdoppenberg hi masters, there's some transposition behavior of Burn Tensor, which makes the grad passed in is also transposed. Hope you could notice it when testing against PyTorch.

For now, it's much quicker to just change PyTorch script with `torch.transpose(linear.weight.data, 0, 1)`

[disccussion on discord](https://discord.com/channels/1038839012602941528/1091796857996451942/1140201837794963559)

<details>
<summary>my new script</summary>

```python
import torch
from torch.nn import Linear
from torch.optim import RMSprop

# Define a Linear layer with given weights and biases


def given_linear_layer(weights: torch.Tensor, biases: torch.Tensor) -> Linear:
    out_features, in_features = weights.shape
    layer = Linear(in_features, out_features, bias=True)

    layer.weight.data = torch.FloatTensor(weights)
    layer.bias.data = torch.FloatTensor(biases)

    return layer


def test_rmsprop():
    # torch.set_printoptions(precision=6, sci_mode=True)
    torch.set_printoptions(precision=6, sci_mode=False)

    LEARNING_RATE = 0.01

    prior_weights = torch.tensor([
        [-0.3206, 0.1374, 0.4043, 0.3200, 0.0859, 0.0671],
        [0.0777, -0.0185, -0.3667, 0.2550, 0.1955, -0.2922],
        [-0.0190, 0.0346, -0.2962, 0.2484, -0.2780, 0.3130],
        [-0.2980, -0.2214, -0.3715, -0.2981, -0.0761, 0.1626],
        [0.3300, -0.2182, 0.3717, -0.1729, 0.3796, -0.0304],
        [-0.0159, -0.0120, 0.1258, 0.1921, 0.0293, 0.3833],
    ])
    prior_bias = torch.tensor(
        [-0.3905, 0.0884, -0.0970, 0.1176, 0.1366, 0.0130])

    linear = given_linear_layer(
        torch.clone(torch.transpose((prior_weights), 0, 1)),
        torch.clone(prior_bias)
    )

    x_1 = torch.tensor([
        [0.6294, 0.0940, 0.8176, 0.8824, 0.5228, 0.4310],
        [0.7152, 0.9559, 0.7893, 0.5684, 0.5939, 0.8883],
    ], requires_grad=True)

    x_2 = torch.tensor([
        [0.8491, 0.2108, 0.8939, 0.4433, 0.5527, 0.2528],
        [0.3270, 0.0412, 0.5538, 0.9605, 0.3195, 0.9085],
    ], requires_grad=True)

    RMSPROP_KWARGS = {
        "lr": LEARNING_RATE,
        "alpha": 0.99,
        "eps": 1e-8,
        "weight_decay": 0.05,
        "momentum": 0.9,
        "centered": False,
    }

    optimizer = RMSprop(
        linear.parameters(),
        **RMSPROP_KWARGS
    )

    # First forward-backward pass and optimization step
    optimizer.zero_grad()
    y = linear(x_1)
    y.backward(torch.ones_like(y))
    optimizer.step()

    # Second forward-backward pass and optimization step
    optimizer.zero_grad()
    y = linear(x_2)
    y.backward(torch.ones_like(y))
    optimizer.step()

    # Print updated state
    print(
        f"\nUpdated layer weights: \n{torch.transpose(linear.weight.data, 0, 1)}")
    # print(
    #     f"\nUpdated layer weights: \n{linear.weight.data}")
    print(f"\nUpdated layer bias: \n{linear.bias.data}")

    # Print difference between prior and updated weights
    print(
        f"\nWeight difference: \n{torch.transpose(linear.weight.data - prior_weights, 0, 1)}")
    # print(
    #     f"\nWeight difference: \n{linear.weight.data - prior_weights}")
    print(f"\nBias difference: \n{linear.bias.data - prior_bias}")
```
</details>

**Comment by antimora** [comment]
@louisfd , @nathanielsimard it seems we completed all the tasks. Can we close it?

**Comment by nathanielsimard** [comment]
Awesome, thanks everyone 🎉
